frame/support/procedural/src/construct_runtime/mod.rs

```diff
@@ -256,7 +259,10 @@ fn construct_runtime_final_expansion(
 	let scrate = generate_crate_access(hidden_crate_name, "frame-support");
 	let scrate_decl = generate_hidden_includes(hidden_crate_name, "frame-support");
 
-	let outer_event = expand::expand_outer_event(&name, &pallets, &scrate)?;
+	let outer_event =
+		expand::expand_outer_enum(&name, &pallets, &scrate, expand::OuterEnumType::Event)?;
+	let outer_error =
+		expand::expand_outer_enum(&name, &pallets, &scrate, expand::OuterEnumType::Error)?;
```

Comment:
Is this regarding the type duplicate?

- in this enum `RuntimeError` we have variants of error type of each pallet
- and each pallet contains:

```rust
pub struct PalletErrorMetadata<T: Form = MetaForm> {
	/// The error type information.
	pub ty: T::Type,
}
```

Or did I construct the `RuntimeError` wrongly here? :D

Reply:
No, I mean that we never return this `RuntimeError` anywhere. Each pallet error is converted to a `DispatchError::Pallet {}`. The type you are storing here will not be able to decode this dispatch error. Or what is your idea here? To convert the `Pallet {}` variant to a `RuntimeError`?

Reply:
Hey, I see now what you mean!

So this `RuntimeError` that I am generating here would effectively be the `ModuleError` from the `DispatchError`:

https://github.com/paritytech/substrate/blob/5b255b2e8ca19d541a944767639265b4d48dcf9a/primitives/runtime/src/lib.rs#L481-L490


Considering that, would you still think that its worth having this outer enum?
Maybe keep it and name it `enum ModuleErrorType`?

Reply:
Just to check my understanding; I've assumed to date that the `u8` index is the pallet index, and then the `[u8; 4]` is the encoding for the pallet `Error` (ie variant index and then whatever fields there might be, normally none). Is that correct?

So yeah, the `enum ModuleErrorType` proposed here would just provide an alternate decode target for the 5 `ModuleError` bytes we get back is my impression.

Reply:
Yes `index` is the index of the pallet in the runtime.

> So yeah, the `enum ModuleErrorType` proposed here would just provide an alternate decode target for the 5 `ModuleError` bytes we get back is my impression.

Yes, but the current form with the runtime level error people will be confused on what this type is or not?

Reply:
For Subxt it would be a net benefit, and let us decode module errors, calls and events in a consistent way (take some bytes, a type ID and a generated type that lines up and decode), so I'm in favour of it for that reason.

But I get that it's not a "real" type, and it isn't quite as obvious as the top level `Call` and `Event` types..

Reply:
Coming back to this, I've kept the `RuntimeError` that is now renamed to `ModuleErrorType` and have implemented the `from_dispatch_error(err: DispatchError) -> Option<Self>` function on it, let me know what you think! 🙏